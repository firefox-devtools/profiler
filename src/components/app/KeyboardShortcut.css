/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.appKeyboardShortcuts {
  position: absolute;
  top: 0;
  left: 0;
  display: none;
  width: 100vw;
  height: 100vh;
  box-sizing: border-box;
  padding: 20px;
  animation: appKeyboardShortcutsFadeIn 0.2s;
  background: #000a;
}

.appKeyboardShortcuts.open {
  display: flex;
  align-items: center;
}

.appKeyboardShortcutsBox {
  overflow: hidden;
  max-width: 800px;

  /* appKeyboardShortcuts margin top + bottom = 40px */
  max-height: calc(100% - 40px);
  border-radius: 5px;
  margin: 0 auto;
  animation: arrowPanelAppear 0.2s cubic-bezier(0.07, 0.95, 0, 1);
  background: #fff;
  filter: drop-shadow(0 0 0.5px rgb(0 0 0 / 0.4))
    drop-shadow(0 4px 5px rgb(0 0 0 / 0.4));
}

.appKeyboardShortcutsScroll {
  overflow-y: scroll;
}

.appKeyboardShortcutsRow {
  display: flex;
  max-width: 400px;
  margin: 7px 0px;
}

.appKeyboardShortcutsLabel {
  flex: 1;
  font-size: 13px;
  padding-inline-end: 25px;
}

.appKeyboardShortcutsShortcut {
  display: inline-flex;
  align-items: center;
  padding: 0 5px;
  border-radius: 3px;
  background: rgb(222 222 227 / 0.79);
  box-shadow: 1px 1px rgb(0 0 0 / 0.27);
  color: #000;
  margin-inline-start: 6px;
}

.appKeyboardShortcutsContent {
  display: grid;
  margin: 40px;
  grid-gap: 80px;
  grid-template-columns: 1fr 1fr;
}

.appKeyboardShortcutsContent h2 {
  margin: 40px 0 10px;
}

.appKeyboardShortcutsContent h2:first-child {
  margin-top: 0;
}

.appKeyboardShortcutsHeader {
  display: flex;
  border-bottom: 1px solid var(--grey-30);
  background: var(--grey-20);
  font-size: 13px;
}

.appKeyboardShortcutsHeaderTitle {
  flex: 1;
  margin: 10px;
}

.appKeyboardShortcutsHeaderClose {
  padding: 6px 6px 6px 30px;
  border: 0;
  border-radius: 3px;

  /* Allow for the photon focus ring to fit in the space by using a 4px margin. */
  margin: 4px;
  background: url(../../../res/img/svg/searchfield-cancel.svg) 10px center
    no-repeat;
  cursor: pointer;
  font-size: inherit;
}

.appKeyboardShortcutsHeaderClose:hover {
  background-color: var(--grey-30);
}

.appKeyboardShortcutsNoInteract {
  pointer-events: none;
}

@keyframes appKeyboardShortcutsFadeIn {
  from {
    background: #0000;
  }

  to {
    background: #000a;
  }
}

/* Do not animate the panel or the background if user prefers reduced motion */
@media (prefers-reduced-motion) {
  .appKeyboardShortcuts,
  .appKeyboardShortcutsBox {
    animation: none;
  }
}

@media (forced-colors: active) {
  .appKeyboardShortcutsBox {
    border: 1px solid CanvasText;
  }

  .appKeyboardShortcutsHeaderClose {
    border: 1px solid ButtonText;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .appKeyboardShortcutsHeaderClose:hover {
    border-color: SelectedItem;
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .appKeyboardShortcutsShortcut {
    background: SelectedItemText;
    color: SelectedItem;
    outline: 1px solid CanvasText;
  }
}
