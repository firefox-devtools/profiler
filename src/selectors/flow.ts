/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
