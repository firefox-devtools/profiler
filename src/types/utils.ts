/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
