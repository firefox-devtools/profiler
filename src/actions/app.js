/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { oneLine } from 'common-tags';
import {
  getSelectedTab,
  getDataSource,
  getLocalTrackOrderByPid,
} from 'firefox-profiler/selectors/url-state';
import {
  getTrackThreadHeights,
  getIsEventDelayTracksEnabled,
  getIsExperimentalCPUGraphsEnabled,
  getIsExperimentalProcessCPUTracksEnabled,
} from 'firefox-profiler/selectors/app';
import {
  getLocalTracksByPid,
  getThreads,
  getCounters,
} from 'firefox-profiler/selectors/profile';
import { sendAnalytics } from 'firefox-profiler/utils/analytics';
import {
  stateFromLocation,
  withHistoryReplaceStateSync,
} from 'firefox-profiler/app-logic/url-handling';
import { finalizeProfileView } from './receive-profile';
import { fatalError } from './errors';
import {
  addEventDelayTracksForThreads,
  initializeLocalTrackOrderByPid,
  addProcessCPUTracksForProcess,
} from 'firefox-profiler/profile-logic/tracks';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getIsCPUUtilizationProvided,
  getAreThereAnyProcessCPUCounters,
} from 'firefox-profiler/selectors/cpu';

import type {
  Profile,
  ThreadsKey,
  CssPixels,
  Action,
  ThunkAction,
  UrlState,
  UploadedProfileInformation,
  IndexIntoCategoryList,
} from 'firefox-profiler/types';
import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';
import type {
  BrowserConnection,
  BrowserConnectionStatus,
} from 'firefox-profiler/app-logic/browser-connection';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function startFetchingProfiles(): Action {
  return { type: 'START_FETCHING_PROFILES' };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when we start setting up the initial url state.
 * It takes the location and profile data, converts the location into url
 * state and then dispatches relevant actions to finalize the view.
 * `profile` parameter can be null when the data source can't provide the profile
 * and the url upgrader step is not needed (e.g. 'from-browser').
 */
export function setupInitialUrlState(
  location: Location,
  profile: Profile | null,
  browserConnection: BrowserConnection | null
): ThunkAction<void> {
  return (dispatch) => {
    let urlState;
    try {
      urlState = stateFromLocation(location, profile);
    } catch (e) {
      if (e.name === 'UrlUpgradeError') {
        // The error is an URL upgrade error, let's fire a fatal error.
        // If there's a service worker update, the class `ServiceWorkerManager`
        // will automatically reload in case the new code knows how to handle
        // this URL version.
        dispatch(fatalError(e));
        return;
      }
      // The location could not be parsed, show a 404 instead.
      console.error(e);
      dispatch(show404(location.pathname + location.search));
      return;
    }

    // Validate the initial URL state. We can't refresh on a from-file URL.
    if (urlState.dataSource === 'from-file') {
      urlState = null;
    }

    // Normally having multiple dispatches is an anti pattern, but here it's
    // necessary because we are doing different things inside those actions and
    // they can't be merged because we are also calling those seperately on
    // other parts of the code.
    // The first dispatch here updates the url state, then changes state as the url
    // setup is done, and lastly finalizes the profile view since everything is set up now.
    // All of this is done while the history is replaced, as this is part of the initial
    // load process.
    withHistoryReplaceStateSync(() => {
      dispatch(updateUrlState(urlState));
      dispatch(finalizeProfileView(browserConnection));
      dispatch(urlSetupDone());
    });
  };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export function reportTrackThreadHeight(
  threadsKey: ThreadsKey,
  height: CssPixels
): ThunkAction<void> {
  return (dispatch, getState) => {
    const trackThreadHeights = getTrackThreadHeights(getState());
    const previousHeight = trackThreadHeights[threadsKey];
    if (previousHeight !== height) {
      // Guard against unnecessary dispatches. This could happen frequently.
      dispatch({
        type: 'UPDATE_TRACK_THREAD_HEIGHT',
        height,
        threadsKey,
      });
    }
  };
}

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}

/**
 * Called when a user has started dragging a file. Used for loading
 * profiles with the drag and drop component.
 */
export function startDragging(): Action {
  return { type: 'START_DRAGGING' };
}

/**
 * Called when a user has stopped dragging a file.
 */
export function stopDragging(): Action {
  return { type: 'STOP_DRAGGING' };
}

/**
 * Called when a custom drag and drop overlay is mounted. This lets
 * the app know that we shouldn't create a default overlay.
 */
export function registerDragAndDropOverlay(): Action {
  return { type: 'REGISTER_DRAG_AND_DROP_OVERLAY' };
}

/**
 * Called when a custom drag and drop overlay is unmounted.
 */
export function unregisterDragAndDropOverlay(): Action {
  return { type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY' };
}

/*
 * This action enables the event delay tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableEventDelayTracks()`
 */
export function enableEventDelayTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsEventDelayTracksEnabled(getState())) {
      console.error(
        'Tried to enable the event delay tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      selectedThreadSelectors.getSamplesTable(getState()).eventDelay ===
      undefined
    ) {
      // Return early if the profile doesn't have eventDelay values.
      console.error(oneLine`
        Tried to enable the event delay tracks, but this profile does
        not have eventDelay values. It is likely an older profile.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addEventDelayTracksForThreads(
      getThreads(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );
    dispatch({
      type: 'ENABLE_EVENT_DELAY_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/*
 * This action enables the CPU graph tracks. They are hidden by default because
 * they are usually for power users and not so meaningful for average users.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableCPUGraphs()`
 */
export function enableExperimentalCPUGraphs(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalCPUGraphsEnabled(getState())) {
      console.error(
        'Tried to enable the CPU graph tracks, but they are already enabled.'
      );
      return false;
    }

    if (!getIsCPUUtilizationProvided(getState())) {
      // Return early if the profile doesn't have threadCPUDelta values.
      console.error(oneLine`
        Tried to enable the CPU graph tracks, but this profile does
        not have threadCPUDelta values. It is likely an older profile.
      `);
      return false;
    }

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_CPU_GRAPHS',
    });

    return true;
  };
}

/*
 * This action enables the process CPU tracks. They are hidden by default because
 * the front-end work is not done for this data yet.
 * There is no UI that triggers this action in the profiler interface. Instead,
 * users have to enable this from the developer console by writing this line:
 * `experimental.enableExperimentalProcessCPUTracks()`
 */
export function enableExperimentalProcessCPUTracks(): ThunkAction<boolean> {
  return (dispatch, getState) => {
    if (getIsExperimentalProcessCPUTracksEnabled(getState())) {
      console.error(
        'Tried to enable the process CPU tracks, but they are already enabled.'
      );
      return false;
    }

    if (
      !getIsCPUUtilizationProvided(getState()) &&
      getAreThereAnyProcessCPUCounters(getState())
    ) {
      // Return early if the profile doesn't have threadCPUDelta values or
      // doesn't have any experimental process CPU counters.
      console.error(oneLine`
        Tried to enable the process CPU tracks, but this profile does
        not have threadCPUDelta values or process CPU threads.
      `);
      return false;
    }

    const oldLocalTracks = getLocalTracksByPid(getState());
    const localTracksByPid = addProcessCPUTracksForProcess(
      getCounters(getState()),
      oldLocalTracks
    );
    const localTrackOrderByPid = initializeLocalTrackOrderByPid(
      getLocalTrackOrderByPid(getState()),
      localTracksByPid,
      null
    );

    dispatch({
      type: 'ENABLE_EXPERIMENTAL_PROCESS_CPU_TRACKS',
      localTracksByPid,
      localTrackOrderByPid,
    });

    return true;
  };
}

/**
 * This caches the profile data in the local state for synchronous access.
 */
export function setCurrentProfileUploadedInformation(
  uploadedProfileInformation: UploadedProfileInformation | null
): Action {
  return {
    type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
    uploadedProfileInformation,
  };
}

export function profileRemotelyDeleted(): Action {
  // Ideally we should store the current profile data in a local indexeddb, and
  // set the URL to /local/<indexeddb-key>.
  return { type: 'PROFILE_REMOTELY_DELETED' };
}

export function updateBrowserConnectionStatus(
  browserConnectionStatus: BrowserConnectionStatus
): Action {
  return { type: 'UPDATE_BROWSER_CONNECTION_STATUS', browserConnectionStatus };
}

export function toggleOpenCategoryInSidebar(
  kind: string,
  category: IndexIntoCategoryList
): Action {
  return {
    type: 'TOGGLE_SIDEBAR_OPEN_CATEGORY',
    kind,
    category,
  };
}
