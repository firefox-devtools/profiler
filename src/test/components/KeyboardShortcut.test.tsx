/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { Provider } from 'react-redux';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import createStore from 'firefox-profiler/app-logic/create-store';
import { KeyboardShortcut } from 'firefox-profiler/components/app/KeyboardShortcut';
import { fireFullKeyPress } from 'firefox-profiler/test/fixtures/utils';
import { coerce } from 'firefox-profiler/utils/flow';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';

describe('app/KeyboardShortcut', function () {
  function setup() {
    const store = createStore();
    const renderResults = render(
      <Provider store={store}>
        <KeyboardShortcut wrapperClassName="exampleClassName">
          <div>Content</div>
          <button type="button">Click me</button>
        </KeyboardShortcut>
      </Provider>
    );

    return { ...renderResults, ...store };
  }

  it('can open and close when hitting ? and Escape respectively', () => {
    const { getByText, queryByText } = setup();

    expect(queryByText('Keyboard shortcuts')).not.toBeInTheDocument();

    // Show the dialog.
    fireFullKeyPress(coerce<Window, HTMLElement>(window), { key: '?' });
    expect(getByText('Keyboard shortcuts')).toBeInTheDocument();

    // Hide the dialog.
    fireFullKeyPress(coerce<Window, HTMLElement>(window), { key: 'Escape' });
    expect(queryByText('Keyboard shortcuts')).not.toBeInTheDocument();
  });

  it('matches the snapshot when closed', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('matches the snapshot when open', () => {
    const { container } = setup();
    fireFullKeyPress(coerce<Window, HTMLElement>(window), { key: '?' });
    expect(container).toMatchSnapshot();
  });

  it('reverts to the previous active element', () => {
    const flushRafCalls = mockRaf();
    const { getByRole } = setup();

    const clickMe = getByRole('button');
    clickMe.focus();

    expect(clickMe).toHaveFocus();

    fireFullKeyPress(coerce<Window, HTMLElement>(window), { key: '?' });
    flushRafCalls(); // The focus is changed after react render.

    expect(getByRole('dialog')).toHaveFocus();

    fireFullKeyPress(coerce<Window, HTMLElement>(window), { key: 'Escape' });
    flushRafCalls(); // The focus is changed after react render.

    expect(clickMe).toHaveFocus();
  });
});
