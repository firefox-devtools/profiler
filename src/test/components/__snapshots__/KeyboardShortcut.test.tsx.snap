// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`app/KeyboardShortcut matches the snapshot when closed 1`] = `
<div>
  <div
    class="exampleClassName"
  >
    <div>
      Content
    </div>
    <button
      type="button"
    >
      Click me
    </button>
  </div>
  <div
    class="appKeyboardShortcuts"
  >
    <div
      aria-labelledby="AppKeyboardShortcutsHeaderTitle"
      aria-modal="true"
      class="appKeyboardShortcutsBox"
      role="dialog"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`app/KeyboardShortcut matches the snapshot when open 1`] = `
<div>
  <div
    class="exampleClassName appKeyboardShortcutsNoInteract"
  >
    <div>
      Content
    </div>
    <button
      type="button"
    >
      Click me
    </button>
  </div>
  <div
    class="appKeyboardShortcuts open"
  >
    <div
      aria-labelledby="AppKeyboardShortcutsHeaderTitle"
      aria-modal="true"
      class="appKeyboardShortcutsBox"
      role="dialog"
      tabindex="0"
    >
      <div
        class="appKeyboardShortcutsHeader"
      >
        <div
          class="appKeyboardShortcutsHeaderTitle"
          id="AppKeyboardShortcutsHeaderTitle"
        >
          Keyboard shortcuts
        </div>
        <button
          class="appKeyboardShortcutsHeaderClose"
          type="button"
        >
          Close
        </button>
      </div>
      <div
        class="appKeyboardShortcutsContent"
      >
        <div
          class="appKeyboardShortcutsColumn"
        >
          <h2>
            Call Tree
          </h2>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Close call node
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowLeft
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Open call node
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowRight
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Open all child call nodes
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              *
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Copy call node label
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ctrl
            </kbd>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              c
            </kbd>
          </div>
          <h2>
            Flame Graph
          </h2>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move view up
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              w
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move view down
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              s
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move selection up
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowUp
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move selection down
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowDown
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move selection left
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowLeft
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Move selection right
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ArrowRight
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Copy call node label
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ctrl
            </kbd>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              c
            </kbd>
          </div>
          <h2>
            Marker Table
          </h2>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Copy marker label
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ctrl
            </kbd>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              c
            </kbd>
          </div>
        </div>
        <div
          class="appKeyboardShortcutsColumn"
        >
          <h2>
            Timeline
          </h2>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Select multiple threads
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              ctrl
            </kbd>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              click
            </kbd>
          </div>
          <h2>
            Call Tree Transforms
          </h2>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Merge function
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              m
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Merge node only
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              M
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Focus on function
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              f
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Focus on subtree only
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              F
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Collapse function
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              c
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Collapse library
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              C
            </kbd>
          </div>
          <div
            class="appKeyboardShortcutsRow"
          >
            <div
              class="appKeyboardShortcutsLabel"
            >
              Drop samples with this function
            </div>
            <kbd
              class="appKeyboardShortcutsShortcut"
            >
              d
            </kbd>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
